### ISSUE TITLE
Dataflow break when using a switch statement with type assertions in golang?

### ISSUE THREAD

**Issue by sud0why** [github/codeql#15350]
Consider the following golang code and codeql dataflow analysis.
The source node `sourceStr` propagate to type assertion like figure 1.
The sink node `sinkStr` reverse propagate to `data` like figure 2.
But there is no data flow from source to sink like figure 3.
Why?

```ql
import go

module UnsafeFlowConfig implements DataFlow::ConfigSig {
  predicate isSource(DataFlow::Node source) {
    source.asExpr().getEnclosingFunction().getName() = "main" and
    source.asExpr().(Ident).getName() = "sourceStr"
  }

  predicate isSink(DataFlow::Node sink) {
    sink.asExpr().getEnclosingFunction().getName() = "main" and
    sink.asExpr().(Ident).getName() = "sinkStr"
  }
}

module UnsafeFlow = TaintTracking::Global<UnsafeFlowConfig>;

// For Debug
int explorationLimit() { result = 30 }

module MyPartialFlowFwd = UnsafeFlow::FlowExplorationFwd<explorationLimit/0>;

predicate adhocPartialFlowFwd(
  string funcLoc, MyPartialFlowFwd::PartialPathNode n, DataFlow::Node src, int dist
) {
  exists(MyPartialFlowFwd::PartialPathNode source |
    MyPartialFlowFwd::partialFlow(source, n, dist) and
    src = source.getNode() and
    funcLoc = n.getNode().getEnclosingCallable().getName()
  )
}

module MyPartialFlowRev = UnsafeFlow::FlowExplorationRev<explorationLimit/0>;

predicate adhocPartialFlowRev(
  string funcLoc, MyPartialFlowRev::PartialPathNode n, DataFlow::Node sinkNode, int dist
) {
  exists(MyPartialFlowRev::PartialPathNode sink |
    MyPartialFlowRev::partialFlow(n, sink, dist) and
    sinkNode = sink.getNode() and
    funcLoc = n.getNode().getEnclosingCallable().getName()
  )
}

from UnsafeFlow::PathNode source, UnsafeFlow::PathNode sink
where UnsafeFlow::flowPath(source, sink)
select sink.getNode(), source, sink, "Unsafe Flow Found."

```

Figure 1:
![17054834295734](https://github.com/github/codeql/assets/16025404/750a551e-5cc6-4266-8634-db25e966c9cc)

Figure 2:
![17054833512722](https://github.com/github/codeql/assets/16025404/b5d88bac-b587-4f10-a6f9-61bd5fc8edf1)

Figure 3:
![170549236324](https://github.com/github/codeql/assets/16025404/75721aab-b797-4dbf-8bb1-877072224e1d)

**Comment by smowton** [comment]
Thanks for the detailed report 👍 I'll take a look.

**Comment by owen-mc** [comment]
Thank you for reporting this. I have spent some time looking into it and the problem seems to be that we don't extract an object (called "entity" in the CodeQL code) for `data` in `data := r.(type)`. I do not yet understand why this is. But the effect is that we don't make a `LocalVariable` and hence an `SsaSourceVariable` for it. SSA is used for data flow, and this is why there is a missing data flow step. I will have to do some more investigation to understand why we do not extract an object for `data` in this case.

**Comment by owen-mc** [comment]
I now fully understand the problem. For type switch statements which define a new variable in the guard (`data := r.(type)` in your example), how the compiler represents things is that this variable is actually a dummy variable, and there is actually a separate variable (with the same name, `data` in your case) for each case clause. I believe this is done so that each case clause gets a variable with the correct type. Our CodeQL library wasn't aware of this, and so we weren't dealing with these implicitly declared variables correctly. I've fixed that in https://github.com/github/codeql/pull/16120, so when it's merged and the next version is released the problem should be fixed.

Thanks again for reporting this! It's always hard to know when aren't representing something correctly and that leads to missing alerts, so it's very valuable to us to get that information and be able to fix the problem.

**Comment by owen-mc** [comment]
https://github.com/github/codeql/pull/16120 has been merged and should be included in release 2.17.1, which is expected in around a week from now. Please reopen this issue if this doesn't solve the problem.
